### PR TITLE
Issue-1076

### PR DESCRIPTION
**Pull request by bell87**
Added defensive programming around incorrectly configured media component to avoid infinite spinners on course page load.

### CODE REVIEW

File: js/adapt-contrib-media.js [hunk 1]
```diff
@@ -81,29 +81,33 @@ define(function(require) {
...
-                Modernizr.load({
-                    load: "assets/froogaloop.js", 
-                    complete: function() {
-                        froogaloopAdded = true;
-                        callback();
-                    }
-                }); 
-                break;
-            default:
+            if (media) {
```
Comment (barry-learningpool) [inline js/adapt-contrib-media.js]: I'd invert this and return early to reduce indentation (only a minor thing)

File: js/adapt-contrib-media.js [hunk 2]
```diff
@@ -81,29 +81,33 @@ define(function(require) {
...
+                    Modernizr.load({
+                        load: "assets/froogaloop.js", 
+                        complete: function() {
+                            froogaloopAdded = true;
+                            callback();
+                        }
+                    }); 
+                    break;
+                default:
+                    callback();
```
Comment (barry-learningpool) [inline js/adapt-contrib-media.js]: Should these be 'return callback()'?

**Comment by barry-learningpool** [comment]
+1

File: js/adapt-contrib-media.js (hunk 2 above)
Comment (moloko) [inline js/adapt-contrib-media.js]: no, `callback` is a pointer to `createPlayer` which doesn't return a value. From the looks of it, the use of `return callback()` on line 98 is there mainly to get the code to exit `addThirdPartyFixes` at that point

**Comment by tomgreenfield** [comment]
I would agree with @barry-learningpool's suggestion on returning early to sink that indentation.

Rather than:
```js
if (!media) {
    callback();
} else {
    // switch statement
}
```
I was expecting:
```js
if (!media) return callback();

// switch statement
```

**Comment by dennis-learningpool** [comment]
+1

**Comment by tomgreenfield** [comment]
+1

**Comment by brian-learningpool** [comment]
+1

**Comment by moloko** [comment]
+2 (tested in Firefox)

Whilst this fixes the 'infinite spinner' issue it does still leave the author with very little indication of what went wrong other than this error:
unhelpful_error

So (and this is just a suggestion), it might be better to do something like this:
```javascript
// We're streaming - set ready now, as success won't be called above
try {
    if (this.model.get('_media').source) {
        this.$('.media-widget').addClass('external-source');
    }
} catch (e) {
    console.error("ERROR! No _media property found in components.json for component " + this.model.get('_id'));
} finally {
    this.setReadyStatus();
}
```
Which gives you this (in Firefox dev tools):
better_error

What do you think? Sorry, I know you've already got all the votes you need to get this merged... so feel free to ignore this if you don't want to have to get everyone to vote again!

**Comment by moloko** [comment]
And finally a small request - with PRs like these it would help anyone who wants to test if you could give a concrete example of what 'incorrectly configured' looks like. Just saves a bit of time as otherwise it's necessary to work it out from the code!

**Comment by bell87** [comment]
Thanks @moloko that is helpful. Always better to give user feedback where possible. hopefully the latest change is ok and I can get this merged.

**Comment by moloko** [comment]
+2 (testing in FF, Chrome and IE11)

File: js/adapt-contrib-media.js [hunk 3]
```diff
@@ -72,39 +72,47 @@ define(function(require) {
                 this.$('audio, video').mediaelementplayer(modelOptions);
 
                 // We're streaming - set ready now, as success won't be called above
-                if (this.model.get('_media').source) {
-                    this.$('.media-widget').addClass('external-source');
+                try {
+                    if (this.model.get('_media').source) {
+                        this.$('.media-widget').addClass('external-source');
+                        this.setReadyStatus();
```
Comment (moloko) [inline js/adapt-contrib-media.js]: this line is redundant due to the `finally` block (anything in a `finally` block will execute regardless of whether there's an error or not)

**Comment by moloko** [comment]
+2

**Pull request merged**